### DISCUSSION TITLE
Append `with` relation callbacks

### DISCUSSION THREAD

**Discussion by jeffreyvr** [laravel/framework#51711] [Q&A]
It might be that I'm just overlooking the obvious, but there (as far as I can tell) doesn't seem to be a way to add multiple callbacks when eager loading relationships.

Example:

```php
Event::first()
  ->attendees()
  ->with('persons', function($query) {
    // does not get executed
    $query->whereFirstName('Taylor');
  })
  ->with('persons', function($query) {
    // gets executed
    $query->orderBy('last_name');
  })
  ->get();
```

Now you might think, just combine it into one callback. But my specific use case comes from custom Sort and Filter classes I have made with [Spatie's Query builder](https://github.com/spatie/laravel-query-builder) package. 

You might create them like so:

```php
class YourCustomFilter implements Filter
{
    public function __invoke($query, $value, $property)
    {
        $query->with(['persons' => function (Builder $query) use ($property, $value) {
            $query->where($property, $value);
        }]);
    }
}

class YourCustomSort implements \Spatie\QueryBuilder\Sorts\Sort
{
    public function __invoke(Builder $query, bool $descending, string $property)
    {
        $query->with(['persons' => function ($query) use ($descending, $property) {
            $query->orderBy($property, $descending ? 'desc' : 'asc');
        }]);
    }
}

// How they are applied in the query builder
QueryBuilder::for($event)
  ->allowedFilters([
    AllowedFilter::custom('filter_name', new YourCustomFilter(), 'property_name'),
  ])
  ->allowedSorts([
    AllowedSort::custom('sort_name', new YourCustomSort(), 'property_name'),
])
```

The above results in that only the callback of the last one added gets executed.

Now I can check within my custom classes if a prior eager loading relationship was already set, and then append the callback's I need. But that feels a little dirty to me. Maybe I'm just overlooking the obvious.

Otherwise, a possible addition to the. Query builder of Laravel might be (just a concept):

```php
/**
 * Set the relationships that should be eager loaded. If already set,
 * merge the callbacks for the specified relationships.
 *
 * @param  string|array  $relations
 * @param  \Closure|null  $callback
 * @return $this
 */
public function withMerge($relations, $callback = null)
{
    foreach ($this->eagerLoad as $name => $constraints) {
        if ($name === $relations) {
            $this->eagerLoad[$name] = function ($query) use ($constraints, $callback) {
                $constraints($query);

                if ($callback instanceof Closure) {
                    $callback($query);
                }
            };

            return $this;
        }
    }

    $this->with($relations, $callback);

    return $this;
}
```

Then you can specifically indicate that you want to append to existing callbacks. I could work on a PR if that is something that is useful. But again, probably overlooking something. 😅

Looking forward to any insights here!

**Comment by wizzymore** [comment]
I don't think it will be what you want anyway, join the relation and add the column you want to sort in the select, this way you can sort by it. If it is what you want good, cause the problem is not there ( maybe ). Are you sure you want the condition on the ->with() and not on the ->whereHas()? Do you really want rows where the condition doesn't apply?  
This is how any spatie filter works by default, it will put the condition on the row not on the relation itself.
